sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java

```diff
@@ -553,14 +554,20 @@ private static ReadableInstant truncateToMillis(Object value) {
     return new Instant((long) value / 1000);
   }
 
-  private static Object convertAvroArray(Field beamField, Object value) {
+  private static Object convertAvroArray(
+      Field beamField, Object value, BigQueryUtils.ConversionOptions options) {
     // Check whether the type of array element is equal.
     List<Object> values = (List<Object>) value;
     List<Object> ret = new ArrayList();
     for (Object v : values) {
-      ret.add(
-          convertAvroPrimitiveTypes(
-              beamField.getType().getCollectionElementType().getTypeName(), v));
+      FieldType arrayElementType = beamField.getType().getCollectionElementType();
+      TypeName elementTypeName = arrayElementType.getTypeName();
+      if (elementTypeName.equals(TypeName.ROW)) {
+        GenericData.Record record = (GenericData.Record) v;
+        ret.add(toBeamRow(record, arrayElementType.getRowSchema(), options));
+      } else {
+        ret.add(convertAvroPrimitiveTypes(elementTypeName, v));
```

Comment:
I think this `if` / `else` piece should just be a recursive call into `convertAvroFormat`. You will need to update it to only need the `FieldType` and not the whole `Field`. This should be easy, I think. I just read the code and it does not seem to actually need the whole `Field`.

Reply:
big +1.

Using `convertAvroFormat` will make sure all other types (not limited to ROW or primitive) can be handled. It that case this is changed to handle `ARRAY<T>`, not limited to `ARRAY<ROW> or ARRAY<primitive type>`

Reply:
The only spot that needed the entire Field was using its name, so unfortunately if we encounter an error we can't inform the user what Field name caused that error (but we can at least tell them the type)